docs: IOx Data Organization and Lifecycle (#3367)
influxdata/influxdb_iox: docs/data_organization_lifecycle.md

> # IOx Data Organization and Lifecycle

Author: Please make sure you select "Display the rich diff` button for review

[^del]: `Deletion` is a large topic that deserves its own document.
[^dup]: The detail of `duplication` and `deduplication` during compaction and query are parts of a large topic that deserve another document.

### Chunk Types
A `Chunk` in IOx is an abstract object defined in the code as a [DbChunk](https://github.com/influxdata/influxdb_iox/blob/12c40b0f0f93e94e483015f9104639a1f766d594/server/src/db/chunk.rs#L78). To optimize the Data LifeCycle and Query Performance, IOx implements these types of physical chunks for a DbChunk: O-MUB, F-MUB, RUB, OS, L-OS.

1. O-MUB: **O**pen **MU**table **B**uffer chunk is optimized for writes and the only chunk type that accepts ingesting data. O-MUB is an in-memory chunk but its data is neither sorted nor encoded.[^type]
1. F-MUB: **F**rozen **MU**table **B**uffer chunk has the same format as O-MUB (in memory, not sorted, not encoded) but it no longer accepts writes. It is used as a transition chunk while its data is being moved from optimized-for-writes to optimized-for-reads.

> 1. RUB: **R**ead **B**uffer chunk is optimized for reads and does not accept writes. RUB is kept in memory and its data is sorted and encoded on the chunk's primary key. Note that since a chunk stores data of a subset of its table columns, Chunk's primary key can also be a subset of its table's primary key.

Reviewer: ```suggestion
1. RUB: **R**ead **B**uffer chunk is optimized for reads and does not accept writes. RUB is kept in memory and its data is sorted and encoded on the chunk's primary key. Note that since a chunk may store data of a subset of its table columns, Chunk's primary key can also be a subset of its table's primary key.
```

1. OS: **O**bject **S**tore chunk is a parquet file of a chunk stored in a durable cloud storage such as Amazon S3 (IOx also supports Azure and Google Clouds). Because an OS is always created from a RUB, it inherits all sorting and encoding properties of the corresponding RUB and Parquet.

> 1. L-OS: **L**ocal-cached **O**bject **S**tore chunk is an OS cached on local non volatile memory of IOx Server. 

Reviewer 2: I'm not sure that we would implement this caching at the in-memory catalog level

[^type]: The detailed format of each chunk type is out of scope of this document

Depending on which stage of the lifecycle a chunk is in, it will be represented by one or a few physical chunk types above.

### Stages of a Chunk

Before digging into Data Lifecycle, let us look into the stages of a chunk implemented as [ChunkStage](https://github.com/influxdata/influxdb_iox/blob/76befe94ad14cd121d6fc5c58aa112997d9e211a/server/src/db/catalog/chunk.rs#L130). A chunk goes through three stages demonstrated in Figure 2: `Open`, `Frozen`, and `Persisted`. 
* When data is ingested into IOx, it will be written into an open chunk which is an `O-MUB`.
* When triggered by some manual or automatic event of the lifecycle (described in next section), the open chunk will be frozen, first to `F-MUB` then transitioned to `RUB`.

> * When the `RUB` is persisted to an `OS` chunk, it stage will be moved to persisted. Unlike the `Open` and `Frozen` stages that are represented by only one type of chunk at a moment in time, the `Persisted` stage can be represented by three chunk types at a time: `RUB`, `OS` and `L-OS` that store the same data for the purpose of query performance. When a query needs to read data of a persisted chunk stage, it will first look for `RUB`, but, if not available, will look for `L-OS`, and then `OS`. `RUB` will be unloaded from the persisted stage if IOx memory runs low, and reloaded if data of that chunk is queried a lot and IOx memory is underused. `L-OS` will be created when data of that chunk is also read lot but no memory available to load its `RUB` back.

Reviewer: I find the notion of `L-OS` confusing. Does it have an actual representation in the code now, and if so what it is? Or perhaps it is planned for a future time?

Or does it refer to the optimized path for parquet files that happen to be stored on a local filesystem backed `ObjectStore`?

Author: I should have mentioned that `L-OS`and `Cache OS` do not exist yet. I think I will remove them from this document. Will update the document if they are available in the future.

Author: removed

└───────────┘       └──────────────────────────┘       │ ┌───────────────┐ │
                                                       │ │L-OS (optional)│ │
                                                       │ └───────────────┘ │
                                                       └───────────────────┘
Figure 2: Stages of a C hunk
```

Now let us see how data of chunks are transformed in IOx's Data LifeCycle.

## Data Life Cycle

IOx Data LifeCycle is a sophisticated combination of **background** and **manual** processes that transform chunks from one stage to another to manage a smooth flow of non-stop ingesting data to their final durable storage while keeping query performance high. The processes are triggered by some events such as `compact`, `delete`, `persist`, `compact OS`, `load RUB`, and `cache OS`. These events are invoked by either users manually or automatically by a `Data LifeCycle Policy`. In this document, we first define the events listed above and the go over examples of those events in Figure 3 to have basic understanding of IOx Data Life Cycle. The `Data LifeCycle Policy` is an advanced topic that will be discussed in future document. 

> ### Data LifeCycle Events

Reviewer: A nitpick but I would probably call these Lifecycle *ACTIONS* rather than Events, to remain consistent with the code (https://github.com/influxdata/influxdb_iox/search?q=ChunkLifecycleAction&type=code).

I also think `Action` better defines what these concepts are, but that is more opinion. Being consistent with the code if possible I think is a much better rationale ;)

Author: Replace all `event` with `action`

> To trigger chunks moving from one stage to another, IOx creates some events. Since IOx keeps chunks inside their (virtual) partition as defined in the previous Section, all chunks described in this section are in the context of one partition. By designed, there is at most one `O-MUB` at a time.

Author: Removed

* **Compact**: is an event to compact eligible `O-MUB`, `F-MUBs`, and `RUBs` into one `RUB`. The eligibility is defined either automatically in the `Data LifeCycle Policy` which is outside the scope of this document or manually specified by users as parameters through that event's gRPC API.
* **Delete**: is an event issued by users (through gRPC API) to delete some data. When a delete is issued, all affected chunks are frozen and its delete predicates are attached to the chunks. So in this Data LifeCycle context, only O-MUBs are affected directly by delete events.

> * **Persist**: is an event to compact eligible `O-MUB`, `F-MUBs`, and `RUBs` into 2 chunks, an `OS` of time window t1-t2 and a `RUB` of time window t2-t3, in which t1 and t3 are the start and end of those chunks' time respectively, and t2 is calculated based on the `Data LifeCycle Policy`. 

Reviewer 2: Possibly worth clarifying this is row timestamps

Author: Done

* **Compact OS**: is an event to compact specified OS chunks into one OS chunk.
* **Load RUB**: is an event to load OS chunks without RUB back in in-memory RUBs.
* **Cache OS**: is an event to cache OS chunks into the IOx server's NVM.

> The `Compact`, `Persist` and `Compact OS` can only compact/persist contiguous chunks which are chunks contain time-window next to each other. They also eliminate permanently duplicated and deleted data. When a new chunk is created, the compacted/persisted ones will be dropped in the same transaction/locking.

Reviewer: ```suggestion
The `Compact`, `Persist` and `Compact OS` can only compact/persist contiguous chunks.  Contiguous chunks contain time-windows of *insert* time (not data timestamp) next to each other. They also eliminate permanently duplicated and deleted data. When a new chunk is created, the compacted/persisted ones will be dropped in the same transaction.
```

### Data LifeCycle Example
Figure 3 illustrates an example of IOx Data LifeCycle with 8 events happening at time T1 to time T8. The number (1, 2, 3, ...) at the end of each chunk name is just for us to identify them easily. They are not the real UUID identifiers of the chunks.

* **Before T1**: There is only `O-MUB 1` in the system to accept eligible writes for its partition.

> * **At T1**: a `Compact` event happens and triggers a process to compact the only `O-MUB 1` into `RUB 1`. In order ot do that, `O-MUB 1` needs to get frozen to `F-MUB 1` first and then compacted into `RUB 1`. If more ingest data comes after T2, a new `O-MUB 2` is created to accept new writes.

Reviewer 2: ```suggestion
* **At T1**: a `Compact` event happens and triggers a process to compact the only `O-MUB 1` into `RUB 1`. In order to do that, `O-MUB 1` needs to get frozen to `F-MUB 1` first and then compacted into `RUB 1`. If more ingest data comes after T2, a new `O-MUB 2` is created to accept new writes.
```

Reviewer 2: Technically, the MUB is closed by the write path which then triggers the compact - but this is an implementation detail I suppose

* **At T2**: a `Delete` is issued that freezes `O-MUB 2` to `F-MUB 2`. Since new data keeps coming, `O-MUB 3` is created to accept new data.
* **At T3**: a `Compact` is requested on `RUB 1`, `F-MUB 2`, and `O-MUB 3`. As always, the `O-MUB 3` must be first frozen to `F-MUB 3` and then compacted with `RUB 1` and `F-MUB 2` to produce `RUB 2`, while `O-MUB 4` is created to accept ingesting data.
* **At T4**: a `Delete` triggers `O-MUB 4` frozen to `F-MUB 4` and `O-MUB 5` created.
* **At T5**: a `Compact` on `F-MUB 4` and `O-MUB 5` to produce `RUB 3` and triggers the creation of `O-MUB 6`.
* **At T6**: a `Persist` is issued on `RUB 2`, `RUB 3` and `O-MUB 6` to produce `OS 1` and `RUB 4`, and triggers the creation of `O-MUB 7`.
* **At T7**: a `Persist` is issued on `RUB 4` and `O-MUB 7` to produce `OS 2` and `RUB 5`, and triggers the creation of `O-MUB 8`.
* **AT T8**: a `Compact OS` is issued on `OS 1` and `OS 2` to produce `OS 3`.

> ```text

Reviewer: this is an epic example
